lib/pharos/yaml_file.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+require 'yaml'
+require 'erb'
+require_relative 'yaml_file/namespace'
+
+module Pharos
+  # Reads YAML files and optionally performs ERB evaluation
+  class YamlFile
+    ParseError = Class.new(StandardError)
+
+    attr_reader :content, :filename
+
+    # @param input [String,IO] A IO/File object, a path to a file or string content
+    # @param override_filename [String] use string as the filename for parse errors
+    # @param force_erb [Boolean] force erb processing even if filename does not end in .yml or is unknown
+    def initialize(input, override_filename: nil, force_erb: false)
+      @filename = override_filename
+      if input.respond_to?(:read)
+        @content = input.read
+        @filename ||= input.respond_to?(:path) ? input.path : nil
+      elsif input.include?("\n")
+        @content = input
+      else
+        @filename ||= input
+        @content = File.read(input)
+      end
+      @force_erb = force_erb
+    end
+
+    def load(variables = {})
+      result = YAML.safe_load(read(variables), [], [], true, @filename)
+      if result.is_a?(String)
+        raise ParseError, "File #{"#{@filename} " if @filename}does not appear to be in YAML format"
+      end
+      result
+    rescue Psych::SyntaxError => ex
+      raise ParseError, ex.message
+    end
+
+    def dirname
+      @filename.nil? ? Dir.pwd : File.dirname(@filename)
+    end
+
+    def basename
+      @filename.nil? ? '<unknown>' : File.basename(@filename)
+    end
+
+    def read(variables = {})
+      erb? ? erb_result(variables) : @content
+    end
+
+    def erb_result(variables = {})
+      Namespace.new(@filename, variables).with_binding do |ns_binding|
+        ERB.new(@content, nil, '%<>-').result(ns_binding)
```

Comment:
Is this missing the `erb.location = [@filename, nil]` for syntax errors, like in #125?

The `rescue` bits add in the filename, but it would be better to get that immediately in the `(erb):121` error with the correct line numbering.

```
(erb):121: syntax error, unexpected end-of-input, expecting keyword_end
t.force_encoding(__ENCODING__)
                              ^ : (erb):121: syntax error, unexpected end-of-input, expecting keyword_end
t.force_encoding(__ENCODING__)
                              ^ in file /app/lib/pharos/resources/weave/daemon-set.yml.erb
/app/lib/pharos/yaml_file.rb:60:in `rescue in erb_result'
/app/lib/pharos/yaml_file.rb:53:in `erb_result'
/app/lib/pharos/yaml_file.rb:50:in `read'
/app/lib/pharos/yaml_file.rb:32:in `load'
/app/lib/pharos/kube.rb:232:in `parse_resource_file'
/app/lib/pharos/kube.rb:129:in `block in apply_stack'
/app/lib/pharos/kube.rb:128:in `each'
/app/lib/pharos/kube.rb:128:in `apply_stack'
/app/lib/pharos/phases/configure_network.rb:50:in `ensure_resources'
/app/lib/pharos/phases/configure_network.rb:25:in `call'
/app/lib/pharos/up_command.rb:131:in `handle_masters'
/app/lib/pharos/up_command.rb:42:in `block in configure'
/app/lib/pharos/up_command.rb:41:in `chdir'
/app/lib/pharos/up_command.rb:41:in `configure'
/app/lib/pharos/up_command.rb:25:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:132:in `run'
/app/lib/pharos/root_command.rb:14:in `run'
bin/pharos-cluster:12:in `<main>'
```

Reply:
```patch
diff --git a/lib/pharos/yaml_file.rb b/lib/pharos/yaml_file.rb
index 2bdb91c..dfceba9 100644
--- a/lib/pharos/yaml_file.rb
+++ b/lib/pharos/yaml_file.rb
@@ -52,12 +52,10 @@ module Pharos
 
     def erb_result(variables = {})
       Namespace.new(@filename, variables).with_binding do |ns_binding|
-        ERB.new(@content, nil, '%<>-').result(ns_binding)
+        erb = ERB.new(@content, nil, '%<>-')
+        erb.location = [@filename, nil]
+        erb.result(ns_binding)
       end
-    rescue NameSpace::Error
-      raise
-    rescue StandardError, ScriptError => ex
-      raise ParseError, "#{ex} : #{ex.message} in file #{@filename}"
     end
 
     private
diff --git a/lib/pharos/yaml_file/namespace.rb b/lib/pharos/yaml_file/namespace.rb
index 16093a8..9866fb8 100644
--- a/lib/pharos/yaml_file/namespace.rb
+++ b/lib/pharos/yaml_file/namespace.rb
@@ -18,8 +18,6 @@ module Pharos
 
       def with_binding(&block)
         yield binding
-      rescue NameError => ex
-        raise Error, "#{ex.message} in file #{@path}"
       end
     end
   end
```